framework/src/executioners/Executioner.C

```diff
@@ -103,3 +110,69 @@ Executioner::addAttributeReporter(const std::string & name, Real initial_value)
 
   return value;
 }
+
+void
+Executioner::addSolveObject(const std::string & type, const std::string & name,
+                            InputParameters & parameters)
+{
+  if (_solve_object_names.count(type))
+    mooseError("Solve object in type ", type, " has been constructed");
+
+  parameters.set<SubProblem *>("_subproblem") = &feProblem();
+  parameters.set<SystemBase *>("_sys") = &feProblem().getNonlinearSystemBase();
```

Comment:
I know it's not the only place they are used, but not a fan of these params added after the fact

Reply:
Not sure if these parameters are absolutely necessary. This is partly copied from `FEProblem::addXXX` functions.

Reply:
Can you pls try and run the test suite? I suspect they are necessary. I dont like them though.

Reply:
I'm pretty sure you can just do `_fe_problem.addObject<SolveObject>(type, name, parameters, false)`.

Reply:
Although, I think I would prefer this code to be in the action. Then you can either have the FlexibleExecutioner access the factory to get the objects, or have a function like `addSolveObject(std::shared_ptr<SolveObject>)` to add it to `_solve_objects`.

Reply:
The relation of SolveObjects and Executioner is like Kernels and FEProblem. Executioner owns the solve objects BUT not fe problem. So calling a function in FEProblem is not proper. We are also not using warehouse for storing solve objects. I should try removing unnecessary parameter settings however.

Reply:
I once attempted to move `FEProblem::solve` into `FEProblemSolve::solve` but not quite successful because things are too entangled. Something in FEProblem should be moved into solve objects.